### ISSUE TITLE
[red-knot] how should `assert_type` handle the float/complex special case?

### ISSUE THREAD

**Issue by carljm** [astral-sh/ruff#16171]
### Description

We follow the typing spec and other type checkers in implementing the special case that a `float` annotation means `int | float`, and a `complex` annotation means `int | float | complex`.

We don't currently follow other type checkers in trying to "hide" this special case by revealing just `float` or `complex` when the actual type we have is `int | float` or `int | float | complex`. Doing this leads to very strange-looking behaviors, like this from pyright:

```py
def _(x: float):
    reveal_type(x)  # float
    if isinstance(x, float):
        reveal_type(x)  # float
    reveal_type(x)  # float | int ??
```

But this leads us to fail this type assertion:

```py
assert_type(3.14, float)
```

Because we interpret `float` there as a type expression, meaning we take it to be `int | float`, but we actually infer just `float` for the literal.
